From 9705669a49c341402efd8528e8fe809379dd798d Mon Sep 17 00:00:00 2001
From: Kazuhito Hagio <k-hagio-ab@nec.com>
Date: Mon, 23 May 2022 14:48:50 +0900
Subject: [PATCH 06/16] Makefile: add missing crash_target.o to be cleaned

Signed-off-by: Kazuhito Hagio <k-hagio-ab@nec.com>
Signed-off-by: Lianbo Jiang <lijiang@redhat.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e520b1217a9d..162c2baa5164 100644
--- a/Makefile
+++ b/Makefile
@@ -184,7 +184,7 @@ GDB_7.6_FILES=
 GDB_7.6_OFILES=${GDB}/gdb/symtab.o
 
 GDB_10.2_FILES=
-GDB_10.2_OFILES=${GDB}/gdb/symtab.o
+GDB_10.2_OFILES=${GDB}/gdb/symtab.o crash_target.o
 
 # 
 # GDB_FLAGS is passed up from the gdb Makefile.
-- 
2.30.2

